clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+ * decode_cache.h: Library that supports caching of instruction decode
+ * information.
+ */
+
+#ifndef _DECODE_CACHE_H_
+#define _DECODE_CACHE_H_ 1
+
+#include "memref.h"
+#include "tracer/raw2trace_shared.h"
+
+#include <cstring>
+#include <mutex>
+#include <unordered_map>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
+ * Base class for storing instruction decode info. Users should sub-class this
+ * base class and implement set_decode_info_derived() to derive and store the decode
+ * info they need.
+ */
+class decode_info_base_t {
+public:
+    /**
+     * Sets the decode info for the provided \p instr which was allocated using the
+     * provided \p dcontext for the provided \p memref_instr. This is done using
+     * the set_decode_info_derived() provided by the derived class. Additionally,
+     * this does other required bookkeeping.
+     */
+    void
+    set_decode_info(void *dcontext,
+                    const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                    instr_t *instr);
+
+    /**
+     * Indicates whether the decode info stored in this object is valid. It won't be
+     * valid if the object is default-constructed without a subsequent
+     * set_decode_info() call. When used with \p decode_cache_t, this indicates
+     * that an invalid instruction was observed at some pc.
+     */
+    bool
+    is_valid() const;
+
+private:
+    /**
+     * Sets the decoding info fields as required by the derived class, based on the
+     * provided instr_t which was allocated using the provided opaque \p dcontext
+     * for the provided \p memref_instr. Derived classes must implement this virtual
+     * function. Note that this cannot be invoked directly as it is private, but
+     * only through set_decode_info() which does other required bookkeeping.
+     *
+     * This is meant for use with \p decode_cache_t, which will invoke
+     * set_decode_info() for each new decoded instruction.
+     *
+     * The responsibility for invoking instr_destroy() on the provided \p instr
+     * lies with this \p decode_info_base_t object, unless
+     * \p decode_cache_t was constructed with \p persist_decoded_instrs_
+     * set to false, in which case no heap allocation takes place.
```

Comment:
But in either case the derived class doesn't need to do anything: so can we replace this whole sentence with just "this base class owns \p instr and is responsible for any deallocation"?

Reply:
> But in either case the derived class doesn't need to do anything

No, when persist_decoded_instrs_ is set, the derived class owns the instr_t and has to instr_destroy() it.